clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -144,13 +159,25 @@ public long metadataExpireMs() {
      */
     public synchronized int requestUpdate() {
         this.needFullUpdate = true;
+        this.backoffUpdateRequests = 0L;
```

Comment:
@AndrewJSchofield & @junrao is my understanding correct?

> For example, if a produce request fails, we request a metadata update and set backoffUpdateRequests to 0

My understanding of the PR is this, metadata request won't backoff, but produce request would backoff. So *likely* metadata is going to be updated next time around produce request is retried(post backoff).

> Otherwise, we likely will still have the stale metadata when retrying the produce request.

So this shouldn't be a problem, correct?

Reply:
> My understanding of the PR is this, metadata request won't backoff, but produce request would backoff. So likely metadata is going to be updated next time around produce request is retried(post backoff).

@msn-tldr : To me, the common reason why a produce request needs to backoff is that the metadata is stable since the latest metadata hasn't been propagated to the brokers yet. So, if we don't backoff the metadata request, the returned metadata may still be stale, which won't help the backed off produce request.

Reply:
The code has been changed to back off metadata requests in situations where the metadata responses are "equivalent", meaning that the epochs for the topic-partitions have not advanced. This is somewhat more sophisticated than was anticipated in the KIP I think, but the intent really was to improve behaviour by introducing backoff, rather than to enumerate every conceivable situation in advance.